### PR TITLE
Improve `bsr @ strided` performance in `baddmm` for `bfloat16/half` with Triton kernels.

### PR DESCRIPTION
**Pull request by nikitaved**
As per title.

Additionally we also introduce support for:
- Rectangular block sizes which are powers of 2 and at least 16 (triton's `dot` limitation).
- Batch support with broadcasting for either of the arguments.

cc @ngimel @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @VitalyFedyunin

### CODE REVIEW

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    # Compute a vector of non-zero elements numbers per each row.
+    # We want to ultimately iterate over non-zero rows.
+    nnz_per_row = crow_indices[:, 1:] - crow_indices[:, :-1]
+
+    # Compute indices of non-zero counts.
+    # batch_idx maps to a broadcasted batch index, while
+    # row_idx tracks non-zero rows of the sparse argument
+    # and rows of the output that get modified.
+    batch_idx, row_idx = nnz_per_row.nonzero(as_tuple=True)
+
```
Comment (nikitaved) [inline torch/sparse/triton_ops/triton_bsr_dense_mm.py:188]: `nonzero` could be removed when the number of nonzero rows is high, then skipping empty rows could be delegated to the kernel.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
+import torch
+import triton
+import triton.language as tl
+
+
+def compressed_indices_to_plain_indices(cidx, pidx):
```
Comment (cpuhrsch) [inline torch/sparse/triton_ops/triton_bsr_dense_mm.py:6]: nit: is this unused?

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    nnz = values.shape[-3]
+    blocksize = values.shape[-2:]
+
+    # Compute broadcasted batch dimension
+    bsr_batch_dims = values.shape[:-3]
+    dense_batch_dims = dense.shape[:-2]
+    batch_dims_broadcasted = torch.broadcast_shapes(bsr_batch_dims, dense_batch_dims)
+
+    # Allocate output
+    output = torch.zeros(
```
Comment (cpuhrsch) [inline torch/sparse/triton_ops/triton_bsr_dense_mm.py:167]: nit: can we use empty?

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    nnz_per_row = crow_indices[:, 1:] - crow_indices[:, :-1]
+
+    # Compute indices of non-zero counts.
+    # batch_idx maps to a broadcasted batch index, while
+    # row_idx tracks non-zero rows of the sparse argument
+    # and rows of the output that get modified.
+    batch_idx, row_idx = nnz_per_row.nonzero(as_tuple=True)
+
+    # Compress the vector of counts to hold only non-zero values.
+    nnz_per_row = nnz_per_row[batch_idx, row_idx]
```
Comment (cpuhrsch) [inline torch/sparse/triton_ops/triton_bsr_dense_mm.py:224]: I wonder what's more expensive. Removing the non-zero counts or skipping them in the kernel.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 1 above)
Comment (cpuhrsch) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:222]: nonzero is also one of those ops that will cause a sync if I'm not mistaken.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    # NOTE: tile_to_blocksize will create a view.
+    # NOTE: output.blocksize[-1] == dense.blocksize[-1],
+    # so it could be any value in [1, dense.shape[-1]).
+    # We need to probably use the largest possible blocksize
+    # so that it fits into SRAM.
+    output = tile_to_blocksize(output, (blocksize[0], blocksize[0]))
+
+    # Compute a vector of non-zero elements numbers per each row.
+    # We want to ultimately iterate over non-zero rows.
+    nnz_per_row = crow_indices[:, 1:] - crow_indices[:, :-1]
```
Comment (cpuhrsch) [inline torch/sparse/triton_ops/triton_bsr_dense_mm.py:215]: I suppose this could be done on the fly within the kernel?

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 4 above)
Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:224]: Depends on how sparse rows are. It is still a draft, but I plan to plant a switch.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 5 above)
Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:215]: Yes, but only if we have plenty of nonzero rows.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+    # batch_idx maps to a broadcasted batch index, while
+    # row_idx tracks non-zero rows of the sparse argument
+    # and rows of the output that get modified.
+    batch_idx, row_idx = nnz_per_row.nonzero(as_tuple=True)
+
+    # Compress the vector of counts to hold only non-zero values.
+    nnz_per_row = nnz_per_row[batch_idx, row_idx]
+    # Compute cumulative counts which along with nnz_per_row
+    # are used to compute offsets into nnz values.
+    nnz_per_row_cumsum = nnz_per_row.cumsum(-1)
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:227]: Yes, but this implementation assumes very sparse inputs for which grid is small, I will update the logic that handles dense inputs, then the look-up table computation could be fused with the mm kernel, for sure.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 5 above)
Comment (cpuhrsch) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:215]: I mean, with modern CUDA you can have a lot of blocks. If you exit them quickly the overhead of a large grid with lots of unused blocks might not be better than the overhead of doing these calculations first.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 2 above)
Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:6]: Not here anymore, right, but it is a very nice util for other kernels we might wanna have for sparse in Triton. We can remove it now...

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 3 above)
Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:167]: No, we cannot, unless we know that each block of output is gonna get hit, otherwise they should be zeroed out.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 5 above)
Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:215]: Sure, makes sense, given that nonzero is  already iterating over nnzs. I will test it out and update benchmarks.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 1 above)
Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:222]: Exactly, so it makes sense to avoid using it in some circumstances which also addresses your comments below.

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 5 above)
Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py]: Introduced a separate kernel for dense rowspaces. I will try to benchmark some reasonable threshold for kernel switch.

File: test/test_sparse_csr.py [hunk 7]
```diff
@@ -1498,6 +1499,51 @@ def run_test_block_addmm_addmv(self,
...
+
+        kernel_run = [False]
+
+        def mm_kernel(*args, **kwargs):
+            kernel_run[0] = True
+            return bsr_dense_mm(*args, **kwargs)
+
+        # Kernel overwrite
+        lib = torch.library.Library("aten", "IMPL")
+        lib.impl("aten::_triton_bsr_dense_mm", mm_kernel, "SparseCsrCUDA")
```
Comment (ngimel) [inline test/test_sparse_csr.py:1526]: wait, so we are registering the kernel in the test only, and it won't be available otherwise?

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+    # Launch kernel
+    n_nnz_block_rows = row_idx.size(-1)
+    n_block_cols = dense.size(-3)
```
Comment (ngimel) [inline torch/sparse/triton_ops/triton_bsr_dense_mm.py:273]: dense.size(-3) can easily be more than 65K

Comment (nikitaved) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:273]: My bad, will push this dim into the x-dim.

File: test/test_sparse_csr.py (hunk 7 above)
Comment (ngimel) [reply test/test_sparse_csr.py:1526]: cc @anjali411, @SherlockNoMad for registering on import

File: torch/sparse/triton_ops/triton_bsr_dense_mm.py (hunk 8 above)
Comment (ngimel) [reply torch/sparse/triton_ops/triton_bsr_dense_mm.py:273]: Is there an easy way to launch the kernel in a loop in this case? That's what's usually done in eager.

File: test/test_sparse_csr.py (hunk 7 above)
Comment (nikitaved) [reply test/test_sparse_csr.py:1527]: @anjali411 , @SherlockNoMad, could you please tell me where it is best to register a kernel overwrite?

Comment (anjali411) [reply test/test_sparse_csr.py:1527]: Hi yeah so I think you wanna add it in `torch/__init__.py` similar to how we add _meta_registrations at import time here: https://github.com/pytorch/pytorch/blob/5faa2792fa3c46f2124d1d1c5f7b6a3865d47d7b/torch/__init__.py#L1136

**Comment by amjames** [comment]
@nikitaved I think my PRs will clear up any of the outstanding numpy dependency check failures. 

However the failures like: 
```
Test results will be stored in test-reports/python-unittest/test_public_bindings
    test_correct_module_names (__main__.TestPublicBindings)
  An API is considered public, if  its  `__module__` starts with `torch.` ... No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
  FAIL (1.951s)
    test_correct_module_names (__main__.TestPublicBindings)
  An API is considered public, if  its  `__module__` starts with `torch.` ...     test_correct_module_names failed - num_retries_left: 3
  Traceback (most recent call last):
    File "/var/lib/jenkins/workspace/test/test_public_bindings.py", line 401, in test_correct_module_names
      self.assertTrue(not failure_list, msg)
  AssertionError: False is not true : All the APIs below do not meet our guidelines for public API from https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation.
  Make sure that everything that is public is expected (in particular that the module has a properly populated `__all__` attribute) and that everything that is supposed to be public does look public (it does not start with `_` and has a `__module__` that is properly populated).
  
  Full list:
  # torch.sparse.triton_ops.triton_bsr_dense_mm.Optional:
    - Is public: it is an attribute that does not start with `_` on a module that does not have `__all__` defined
    - Does NOT look public: because its `__module__` attribute (`typing`) is not within the torch library or does not start with the submodule where it is defined (`torch.sparse.triton_ops.triton_bsr_dense_mm`)
    - You can do either of these two things to fix this problem:
      - To make it NOT public: either define a `__all__` for `torch.sparse.triton_ops.triton_bsr_dense_mm` or add a `_` at the beginning of the name
      - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.sparse.triton_ops.triton_bsr_dense_mm`
  # torch.sparse.triton_ops.triton_bsr_dense_mm.Tuple:
    - Is public: it is an attribute that does not start with `_` on a module that does not have `__all__` defined
    - Does NOT look public: because its `__module__` attribute (`typing`) is not within the torch library or does not start with the submodule where it is defined (`torch.sparse.triton_ops.triton_bsr_dense_mm`)
    - You can do either of these two things to fix this problem:
      - To make it NOT public: either define a `__all__` for `torch.sparse.triton_ops.triton_bsr_dense_mm` or add a `_` at the beginning of the name
      - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.sparse.triton_ops.triton_bsr_dense_mm`
  # torch.sparse.triton_ops.triton_bsr_dense_mm.has_triton:
    - Is public: it is an attribute that does not start with `_` on a module that does not have `__all__` defined
    - Does NOT look public: because its `__module__` attribute (`torch._inductor.utils`) is not within the torch library or does not start with the submodule where it is defined (`torch.sparse.triton_ops.triton_bsr_dense_mm`)
    - You can do either of these two things to fix this problem:
      - To make it NOT public: either define a `__all__` for `torch.sparse.triton_ops.triton_bsr_dense_mm` or add a `_` at the beginning of the name
      - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.sparse.triton_ops.triton_bsr_dense_mm`
```

Are legitimate and need to be addressed. The advice in the error message should suffice, either the module needs to be private, or a `__all__` should be added to prevent leaking non-library or private module code into the public API.

File: torch/sparse/triton_ops/__init__.py [hunk 9]
```diff
@@ -1 +1,5 @@
 from .triton_bsr_dense_mm import bsr_dense_mm
+
+__all__ = [
+    "bsr_dense_mm",
```
Comment (amjames) [inline torch/sparse/triton_ops/__init__.py:4]: 🚀

Comment (cpuhrsch) [reply torch/sparse/triton_ops/__init__.py:4]: > 🚀

🚀

**Comment by cpuhrsch** [comment]
@nikitaved @amjames - Is this kind of integration still feasible or should we spend that time teaching torch.compile how to emit this code?

**Comment by nikitaved** [comment]
@cpuhrsch , the CI failures are real. These happen at import time with `from torch._dynamo.utils import has_triton`.

**Comment by cpuhrsch** [comment]
@nikitaved - Are they locally reproducible? Do you have a sense for the root cause?

**Comment by nikitaved** [comment]
Yes, they are reproducible. I have not yet figured out the root cause.

**Comment by nikitaved** [comment]
@amjames , the ROCM issue seemed real, I have rebased the PR, let's see what it shows now..

**Comment by amjames** [comment]
@nikitaved ROCM support is still WIP for triton, some things are working apparently but I would say it is a beta feature at this point and we might not be able to rely on it.

They have a tracking issue up https://github.com/openai/triton/issues/252

**Comment by cpuhrsch** [comment]
We don't need to support ROCM for this PR. Let's disable ROCM support at a C++ level and skip the tests.

**Comment by nikitaved** [comment]
Fingers crossed.

**Comment by nikitaved** [comment]
@cpuhrsch , @amjames , it appears like it is good to go :) The lint issues are irrelevant.

**Review by cpuhrsch — APPROVED** [review]

**Comment by malfet** [comment]
@nikitaved why was it landed, as it clearly breaks CI?

**Comment by amjames** [comment]
@malfet the linter is signaling on a file that the PR didn't touch. I don't know why that is happening, but in theory the version of the problematic file is the same before or after the PR?

**Comment by cpuhrsch** [comment]
Error seems real
```
2023-01-18T12:59:03.4340131Z   test_triton_bsr_dense_bmm_block_size_16_int32_cuda_bfloat16 (__main__.TestSparseCSRCUDA) ... [248872bd2a50:1165 :0:1165] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x7ffced1deff8)
2023-01-18T12:59:03.4340221Z ==== backtrace (tid:   1165) ====
2023-01-18T12:59:03.4340349Z  0  /usr/lib/libucs.so.0(ucs_handle_error+0x144) [0x7efba458b2f4]
2023-01-18T12:59:03.4340465Z  1  /usr/lib/libucs.so.0(+0x3666c) [0x7efba458b66c]
2023-01-18T12:59:03.4340575Z  2  /usr/lib/libucs.so.0(+0x368e4) [0x7efba458b8e4]
2023-01-18T12:59:03.4340811Z  3  /lib/x86_64-linux-gnu/libpthread.so.0(+0x12980) [0x7efc03b01980]
2023-01-18T12:59:03.4341073Z  4  /opt/conda/bin/../lib/libstdc++.so.6(_ZNK10__cxxabiv120__si_class_type_info12__do_dyncastElNS_17__class_type_info10__sub_kindEPKS1_PKvS4_S6_RNS1_16__dyncast_resultE+0x10) [0x7efbe2da6160]
```
for
```
[linux-bionic-cuda11.6-py3.10-gcc7-sm86 / test (default, 4, 4, linux.g5.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/3948211284/jobs/6758971244#logs)
```
which is the sm86 (e.g. A100) specific GPU test.

@nikitaved does this reproduce locally?

**Comment by nikitaved** [comment]
@cpuhrsch , this is the very first time I see such an error, it was always green.

File: torch/cuda/__init__.py [hunk 10]
```diff
@@ -898,18 +898,18 @@ def _dtype(self):
...
+torch._storage_classes.add(LongStorage)  # type: ignore[has-type]
+torch._storage_classes.add(IntStorage)  # type: ignore[has-type]
+torch._storage_classes.add(ShortStorage)  # type: ignore[has-type]
+torch._storage_classes.add(CharStorage)  # type: ignore[has-type]
+torch._storage_classes.add(ByteStorage)  # type: ignore[has-type]
+torch._storage_classes.add(HalfStorage)  # type: ignore[has-type]
+torch._storage_classes.add(BoolStorage)  # type: ignore[has-type]
+torch._storage_classes.add(BFloat16Storage)  # type: ignore[has-type]
+torch._storage_classes.add(ComplexDoubleStorage)  # type: ignore[has-type]
+torch._storage_classes.add(ComplexFloatStorage)  # type: ignore[has-type]
```
Comment (malfet) [inline torch/cuda/__init__.py:912]: I'm really confused, why this change was necessary?

**Comment by cpuhrsch** [comment]
@nikitaved @amjames - looks like the flaky failure might be real. Let's run the sparse_csr test in a loop and see if we can get it to fail locally.

**Comment by malfet** [comment]
Well, revert indeed resulted in 3rd shard being stable again, so this PR was a culprit, but somehow it either deadlocked or OOMed the runners so I don't have a log of what was going on

File: torch/cuda/__init__.py (hunk 10 above)
Comment (nikitaved) [reply torch/cuda/__init__.py:912]: @amjames , could it be there is a leakage coming from the lazy initialization process?

**Comment by nikitaved** [comment]
@cpuhrsch , the inputs are deterministic. I still cannot reproduce anything locally. Will try to log in into the machine.

File: torch/cuda/__init__.py (hunk 10 above)
Comment (nikitaved) [reply torch/cuda/__init__.py:912]: And now we have even more coming our way...

**Comment by nikitaved** [comment]
@pytorchbot merge -g.

Follow-up: resolve import.

**Comment by nikitaved** [comment]
@pytorchbot merge -g

Follow-up: resolve import.

**Comment by nikitaved** [comment]
I confirm, the cancelled job deadlocks.

**Comment by ZainRizvi** [comment]
@pytorchbot revert -m "Causing a test to fail consistently: test_decomp.py::HasDecompTest::test_has_decomposition" -c landrace

See https://hud.pytorch.org/pytorch/pytorch/commit/46f16b93636615a81242b0d5cded84c5a57fd2e2 for more details

**Comment by malfet** [comment]
@nikitaved it would be good to update the PR description to explain what was fixed since the last revert.
Also, I wonder why "revertedx2" label is not there..

**Comment by nikitaved** [comment]
@malfet , nothing was really fixed, the kernel was just hanging there, but new tests showed up that scan native functions including dummy entries to be overwritten with Triton implementations.

File: test/test_sparse_csr.py [hunk 11]
```diff
@@ -20,6 +20,7 @@
     floating_types, all_types_and_complex_and, floating_and_complex_types, floating_types_and,
     all_types_and_complex, floating_and_complex_types_and
 )
+from torch._inductor.utils import has_triton
```
Comment (cpuhrsch) [inline test/test_sparse_csr.py:23]: Let's try moving this into test local scope

**Comment by nikitaved** [comment]
From the offline discussion with @cpuhrsch : we decided to remove all the Cpp hooks for now. `test_decomp` issues are real, but they manifest themselves after a very long time spinning the tests. We will try to investigate these issues in a follow-up PR.

**Comment by nikitaved** [comment]
All right, CUDA 11.6 is deprecated. Let's spin this one again.

**Comment by nikitaved** [comment]
Closing in favor of https://github.com/pytorch/pytorch/pull/94823 for more granular issue control.
